File: emm/server_service/src/schemas/acs_settings.py
"""Pydantic-схемы настроек доступа к ACS (`/settings/acs`, `/settings/acs/departments`).

Платформенный singleton под `account_admin` + per-department opt-in таблица.
Пароль clonezilla-сервера никогда не возвращается в открытом виде — только
факт "задан/не задан" (`password_is_set`); обновление идёт write-only полем
в PUT.
"""

from datetime import datetime

from pydantic import BaseModel, Field


class AcsSettingsResponse(BaseModel):
    """Текущие настройки доступа к ACS."""

    enabled: bool = Field(description="Общий кил-свитч снимков ACS на всю платформу.")
    acs_url: str | None = Field(default=None, description="Базовый URL ACS.")
    password_is_set: bool = Field(description="Задан ли пароль clonezilla-сервера. Само значение не отдаётся.")


class AcsSettingsUpdate(BaseModel):
    """Тело PUT — частичное обновление настроек ACS.

    Любое поле можно опустить — тогда текущее значение сохраняется.
    `acs_password` — plaintext на вход, шифруется на сервисном слое перед
    сохранением; пустое/отсутствующее значение не трогает уже сохранённый
    пароль. Чтобы явно стереть пароль (например, при смене clonezilla-сервера
    без немедленной замены креда) — `clear_password=True`.
    """

    enabled: bool | None = Field(default=None, description="Включить/выключить снимки ACS.")
    acs_url: str | None = Field(default=None, max_length=512, description="Базовый URL ACS.")
    acs_password: str | None = Field(
        default=None,
        description="Новый пароль clonezilla-сервера (plaintext). Пусто — не менять текущий.",
    )
    clear_password: bool = Field(
        default=False,
        description="Явно стереть сохранённый пароль (игнорируется, если одновременно передан `acs_password`).",
    )


class AcsDepartmentAccessItem(BaseModel):
    """Одна строка списка отделов с флагом доступа к снимкам ACS."""

    department_id: str = Field(description="ID отдела (auth_service).")
    is_enabled: bool = Field(description="Включён ли отделу доступ к снимкам ACS.")
    updated_at: datetime | None = Field(
        default=None, description="Когда флаг менялся в последний раз. None — отдел ещё ни разу не настраивался."
    )
    created_by: str | None = Field(default=None, description="Кто впервые включил/выключил отделу доступ.")


class AcsDepartmentAccessListResponse(BaseModel):
    """Ответ GET /settings/acs/departments."""

    items: list[AcsDepartmentAccessItem]


class AcsDepartmentAccessUpdateItem(BaseModel):
    """Один апдейт в батче PUT /settings/acs/departments."""

    department_id: str = Field(description="ID отдела, которому меняем флаг.")
    is_enabled: bool = Field(description="Новое значение флага доступа.")


class AcsDepartmentAccessUpdate(BaseModel):
    """Тело PUT /settings/acs/departments — один или несколько флагов за раз."""

    items: list[AcsDepartmentAccessUpdateItem] = Field(min_length=1, max_length=200)


class AcsInternalSettingsResponse(BaseModel):
    """Ответ GET /internal/settings/acs — для server_worker."""

    acs_url: str = Field(description="Базовый URL ACS.")
    acs_password: str = Field(description="Расшифрованный пароль clonezilla-сервера (plaintext, только worker'у).")

File: emm/web_ui/src/api/server/acsSettings.ts
/**
 * Настройки доступа к ACS — внешнему сервису снимков дисков физических
 * серверов (обёртка над Clonezilla).
 *
 * Платформенный singleton (`enabled`/`acs_url`/пароль) + per-department
 * opt-in таблица поверх обычной action-матрицы. Пароль clonezilla-сервера
 * write-only — GET отдаёт только факт `password_is_set`, значение никогда
 * не возвращается.
 *
 * Источник истины — `server_service` endpoint'ы:
 *   GET  /api/server/v1/settings/acs
 *   PUT  /api/server/v1/settings/acs
 *   GET  /api/server/v1/settings/acs/departments
 *   PUT  /api/server/v1/settings/acs/departments
 * Гейтится account_admin; остальным backend ответит 403.
 *
 * `/settings/acs/departments` не несёт имён отделов — server_service своей
 * таблицы departments не хранит. Имена подтягиваются на фронте отдельно
 * через `listDepartments()` (`@/api/auth/departments`) и сшиваются по
 * `department_id`.
 */

import { apiGet, apiPut } from "@/api/client";

/** Текущие настройки ACS (ответ GET /settings/acs). */
export interface AcsSettings {
  /** Общий кил-свитч снимков ACS на всю платформу. */
  enabled: boolean;
  /** Базовый URL ACS. */
  acs_url: string | null;
  /** Задан ли пароль clonezilla-сервера. Само значение не отдаётся. */
  password_is_set: boolean;
}

/** Тело PUT /settings/acs — частичное обновление. */
export interface AcsSettingsUpdate {
  enabled?: boolean;
  acs_url?: string;
  /** Новый пароль, plaintext. Пусто/не передано — не менять текущий. */
  acs_password?: string;
  /** Явно стереть сохранённый пароль (игнорируется вместе с acs_password). */
  clear_password?: boolean;
}

/** Один отдел из GET /settings/acs/departments — без имени, только флаг. */
export interface AcsDepartmentAccessItem {
  department_id: string;
  is_enabled: boolean;
  updated_at: string | null;
  created_by: string | null;
}

/** Один апдейт в батче PUT /settings/acs/departments. */
export interface AcsDepartmentAccessUpdateItem {
  department_id: string;
  is_enabled: boolean;
}

const BASE = "/server/v1";

/** Прочитать текущие настройки ACS. 403 — нет account_admin. */
export function getAcsSettings(): Promise<AcsSettings> {
  return apiGet<AcsSettings>(`${BASE}/settings/acs`);
}

/** Обновить настройки ACS (частичная замена). */
export function updateAcsSettings(
  body: AcsSettingsUpdate,
): Promise<AcsSettings> {
  return apiPut<AcsSettings>(`${BASE}/settings/acs`, body);
}

/** Прочитать список отделов с текущим флагом доступа к снимкам ACS. */
export async function getAcsDepartmentAccess(): Promise<
  AcsDepartmentAccessItem[]
> {
  const res = await apiGet<{ items: AcsDepartmentAccessItem[] }>(
    `${BASE}/settings/acs/departments`,
  );
  return res.items;
}

/** Переключить флаг(и) доступа отдела(ов) к снимкам ACS за один запрос. */
export async function updateAcsDepartmentAccess(
  items: AcsDepartmentAccessUpdateItem[],
): Promise<AcsDepartmentAccessItem[]> {
  const res = await apiPut<{ items: AcsDepartmentAccessItem[] }>(
    `${BASE}/settings/acs/departments`,
    { items },
  );
  return res.items;
}
